test/transform/forEachTest.js

```diff
@@ -120,98 +107,197 @@ describe('For loops to for-each', () => {
         ).toReturn(
-          'array.forEach((item, i) => {\n' +
-          '  console.log(item, i);\n' +
+          'xs.forEach(x => {\n' +
+          '  console.log(x);\n' +
           '});'
         );
       });
 
-      it('inside a function in loop body', () => {
+      it('for nested loops', () => {
         expectTransform(
-          'for (let i=0; i < array.length; i++) {\n' +
-          '  let item = array[i];\n' +
-          '  callback(function(){ return i; });\n' +
+          'for (let i = 0; i < xs.length; i++) {\n' +
+          '  const x = xs[i];\n' +
+          '  console.log(xs[i]);\n' +
+          '  for (let j = 0; j < xs[i].length; j++) {\n' +
+          '    const y = xs[i][j];\n' +
+          '    console.log(xs[i][j]);\n' +
+          '  }\n' +
           '}'
         ).toReturn(
-          'array.forEach((item, i) => {\n' +
-          '  callback(function(){ return i; });\n' +
+          'xs.forEach(x => {\n' +
+          '  console.log(x);\n' +
+          '\n' +
+          '  x.forEach(y => {\n' +
+          '    console.log(y);\n' +
+          '  });\n' +
           '});'
         );
       });
     });
 
-    describe('should transform array items used inside the body', () => {
-      it('for simple loops', () => {
+    it('should transform array references that do not modify', () => {
+      expectTransform(
+        'for (let i = 0; i < xs.length; i++) {\n' +
+        '  const x = xs[i];\n' +
+        '  a = xs[i];\n' +
+        '  b = xs[i] + 1;\n' +
+        '  c = f(xs[i]);\n' +
+        '  d[xs[i]] = 1;\n' +
+        '}'
+      ).toReturn(
+        'xs.forEach(x => {\n' +
+        '  a = x;\n' +
+        '  b = x + 1;\n' +
+        '  c = f(x);\n' +
+        '  d[x] = 1;\n' +
+        '});'
+      );
+    });
+
+    describe('should not transform array references to use the alias', () => {
+      it('when assigning to the array reference', () => {
         expectTransform(
           'for (let i = 0; i < xs.length; i++) {\n' +
           '  const x = xs[i];\n' +
-          '  console.log(xs[i]);\n' +
+          '  xs[i] = 1;\n' +
+          '}'
+        ).toReturn(
+          'xs.forEach((x, i) => {\n' +
+          '  xs[i] = 1;\n' +
+          '});'
+        );
+      });
+
+      it('when modifying the array reference with an update', () => {
+        expectTransform(
+          'for (let i = 0; i < xs.length; i++) {\n' +
+          '  const x = xs[i];\n' +
+          '  xs[i]++;\n' +
+          '}'
+        ).toReturn(
+          'xs.forEach((x, i) => {\n' +
+          '  xs[i]++;\n' +
+          '});'
+        );
+      });
+
+      it('when modifying the array reference with an update expression', () => {
+        expectTransform(
+          'for (let i = 0; i < xs.length; i++) {\n' +
+          '  const x = xs[i];\n' +
+          '  xs[i] -= 1;\n' +
+          '}'
+        ).toReturn(
+          'xs.forEach((x, i) => {\n' +
+          '  xs[i] -= 1;\n' +
+          '});'
+        );
+      });
+    });
```

Comment:
Argh, that is a good catch. The frustrating bit is that I've found the above transformation very useful personally when it comes to refactoring old code, but I agree that it's not easy to get correct. I think I might have to exclude anything with a function call inside :(